src/gnu/prolog/database/Module.java

```diff
@@ -183,4 +262,150 @@ public void removePredicateListener(PredicateListener listener)
+
+	/**
+	 * A {@link WeakReference} to a {@link PrologCodeListener}
+	 * 
+	 */
+	private static class PrologCodeListenerRef extends WeakReference<PrologCodeListener>
+	{
+		PrologCodeListenerRef(ReferenceQueue<? super PrologCodeListener> queue, PrologCodeListener listener,
+				CompoundTermTag tag)
+		{
+			super(listener, queue);
+			this.tag = tag;
+		}
+
+		CompoundTermTag tag;
+	}
+
+
+	/**
+	 * add prolog code listener
+	 * 
+	 * @param tag
+	 * @param listener
+	 */
+	public void addPrologCodeListener(Environment env, CompoundTermTag tag, PrologCodeListener listener)
+	{
+		synchronized (tag2listeners)
+		{
+			pollPrologCodeListeners();
+			List<PrologCodeListenerRef> list = tag2listeners.get(tag);
+			if (list == null)
+			{
+				list = new ArrayList<PrologCodeListenerRef>();
+				tag2listeners.put(tag, list);
+			}
+			list.add(new PrologCodeListenerRef(prologCodeListenerReferenceQueue, listener, tag));
+		}
+	}
+
+	/**
+	 * remove prolog code listener
+	 *
+	 * @param env
+	 * @param tag
+	 * @param listener
+	 */
+	public void removePrologCodeListener(Environment env, CompoundTermTag tag, PrologCodeListener listener)
+	{
+		synchronized (tag2listeners)
+		{
+			pollPrologCodeListeners();
+			List<PrologCodeListenerRef> list = tag2listeners.get(tag);
+			if (list != null)
+			{
+				ListIterator<PrologCodeListenerRef> i = list.listIterator();
+				while (i.hasNext())
+				{
+					PrologCodeListenerRef ref = i.next();
+					PrologCodeListener lst = ref.get();
+					if (lst == null)
+					{
+						i.remove();
+					}
+					else if (lst == listener)
+					{
+						i.remove();
+						return;
+					}
+				}
+			}
+		}
+	}
+
+	public void predicateUpdated(Environment env, PredicateUpdatedEvent evt)
+	{
+		PrologCode code = tag2code.remove(evt.getTag());
+		pollPrologCodeListeners();
+		if (code == null) // if code was not loaded yet
+		{
+			return;
+		}
+		CompoundTermTag tag = evt.getTag();
+		synchronized (tag2listeners)
+		{
+			List<PrologCodeListenerRef> list = tag2listeners.get(tag);
+			if (list != null)
+			{
+				PrologCodeUpdatedEvent uevt = new PrologCodeUpdatedEvent(env, tag);
+				ListIterator<PrologCodeListenerRef> i = list.listIterator();
+				while (i.hasNext())
+				{
+					PrologCodeListenerRef ref = i.next();
+					PrologCodeListener lst = ref.get();
+					if (lst == null)
+					{
+						i.remove();
+					}
+					else
+					{
+						lst.prologCodeUpdated(uevt);
+						return;
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * get prolog code
+	 * 
+	 * @param tag
+	 * @return the {@link PrologCode} for the tag
+	 * @throws PrologException
+	 */
+	public synchronized PrologCode getPrologCode(Environment env, CompoundTermTag tag) throws PrologException
+	{
+		PrologCode code = tag2code.get(tag);
+		if (code == null)
+		{
+			code = env.loadPrologCode(tag);
+			tag2code.put(tag, code);
+		}
+		return code;
+
+	}
+
+	protected void pollPrologCodeListeners()
```

Comment:
Yep. I think you added it with 1fb588a72b1176d0b0f42741d494294eb211de59. I'll add some sort of comment though